File: padmavati_crm/public/py/customer.py
import frappe

@frappe.whitelist()
def generate_unique_code(custom_location):
    
    location = custom_location
    code_parts = []

    # Traverse the location hierarchy to build the code
    while location:
        loc_doc = frappe.get_doc('Geographic Location', location)
        code_parts.insert(0, str(loc_doc.location_code))  # Convert the integer to a string
        if not loc_doc.parent_location:
            break  # Break if the top node is reached
        location = loc_doc.parent_location

    print("\n\n\n\nCode parts:", code_parts)

    # Combine the parts to form the unique employee code
    employee_code = ''.join(code_parts)
    # employee_code += "- EMP"
    # employee_code += "-{:03d}".format(1)  # Assuming a default or initial value of 1

    print("\n\n\n\nEmployee Code:", employee_code)
    return employee_code



@frappe.whitelist()
def create_user(lead_name):
    # Fetch the Lead document
    lead = frappe.get_doc("Lead", lead_name)

    # Get the email and first name from the Lead
    email = lead.email_id
    full_name = lead.first_name

    # Check if a User with this email already exists
    if not frappe.db.exists("User", email):
        # Create a new User document
        user = frappe.get_doc({
            "doctype": "User",
            "email": email,
            "first_name": full_name,
            "enabled": 1,
            "send_welcome_email": 1
        })
        user.insert(ignore_permissions=True)
        frappe.db.commit()
        return {"message": "User created successfully!"}
    else:
        return {"message": "User already exists with this email."}

File: padmavati_crm/public/js/customer.js
frappe.ui.form.on('Customer', {
    before_save: function (frm) {
        if (frm.doc.custom_location) {
            let args = {
                custom_location: frm.doc.custom_location,
            };

            frappe.call({
                method: 'padmavati_crm.public.py.customer.generate_unique_code',
                args: args,
                callback: function (response) {
                    if (response.message) {
                        console.log(response.message);
                        frm.set_value('custom_unique_id', response.message);
                    }
                }
            });
        } else {
            frappe.msgprint(__('To generate a unique code for the employee, the location is required.'));
            frappe.validated = false;
        }
    },
    refresh: function(frm) {
        console.log(frm.doc.lead_name);
        // Add custom button "Create User"
        frm.add_custom_button(__('Create User'), function() {
            // Call the server-side method to create the user
            frappe.call({
                method: 'padmavati_crm.public.py.customer.create_user',
                args: {
                    lead_name: frm.doc.lead_name  // Pass the lead name as argument
                },
                callback: function(response) {
                    console.log(response);
                    if (response.message === "User created successfully!") {
                        frappe.msgprint(__('User created successfully!'));
                    } else {
                        frappe.msgprint(__('User already exists with this email.'));
                    }   
                }
            });
        });
    }

});
